salt/cli/batch_async.py

```diff
@@ -221,7 +221,6 @@ def end_batch(self):
                 "metadata": self.metadata
             }
             self.event.fire_event(data, "salt/batch/{0}/done".format(self.batch_jid))
-            self.event.remove_event_handler(self.__event_handler)
```

Comment:
@dincamihai curiosity: who/what used the event handler after `end_batch` and before `__del__`?

Reply:
something in thin call self.event.remove_event_handler(self.__event_handler)
so i've moved it in __del__

Reply:
Not sure I understand. Are you saying that `remove_event_handler` uses the event handler itself?

Reply:
I have not actually investigated more after I've observed that moving that line fixed the problem.

What I think it happens is that we probably call `remove_event_handler` too soon.
It could be that when we call it, this `__del__` would be called: https://github.com/openSUSE/salt/blob/6c84612b52b5f14e74a1c44f03d78a85c6f0c5dc/salt/transport/ipc.py#L367 which would call https://github.com/openSUSE/salt/blob/6c84612b52b5f14e74a1c44f03d78a85c6f0c5dc/salt/transport/ipc.py#L384 which would probably trigger more cleanup.

I can see that `fire_event` should happen before `remove_event_handler`, but since `fire_event` is spawning a callback https://github.com/openSUSE/salt/blob/189a19b6e8d28cc49e5ad5f2a683e1dfdce66a86/salt/utils/event.py#L756 it could still be that `remove_event_handler` is called before `fire_event`. Maybe even before the `fire_event` for batch-start.

I know this explanation has some amount of assumptions, if you want me to be more pricese I would have to do more debugging :)

Let me know if it makes sense.

Reply:
It does make sufficient sense for the problem at hand.

The only thing I would recommend (only if you have some time) is to repeat the API stress test to make sure we aren't leaking any after this change.

Reply:
Ok. I can do that. I will post back the outcome.

Reply:
![image](https://user-images.githubusercontent.com/1615643/69882067-44cc6e80-12cf-11ea-91dd-6a32bb9a10f6.png)

I executed the stress test with this patch and the memory leak is not present.
Will merge and promote on Monday.

Reply:
:+1: